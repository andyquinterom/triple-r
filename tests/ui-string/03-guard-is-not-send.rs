use std::thread;
use triple_r::ReusableString;

fn main() {
    let mut s = ReusableString::default();
    let guard = s.recycle();

    // This closure requires `guard` to be `Send`.
    let handle = thread::spawn(move || {
        // The compiler should prevent this because `ReusableStringGuard` is not `Send`.
        // It holds a raw pointer to the string's data, which is not safe to move across
        // threads.
        println!("Length: {}", guard.len());
    });

    handle.join().unwrap();
}
