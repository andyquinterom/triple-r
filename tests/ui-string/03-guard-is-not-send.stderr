error[E0277]: `*mut String` cannot be sent between threads safely
  --> tests/ui-string/03-guard-is-not-send.rs:9:32
   |
9  |       let handle = thread::spawn(move || {
   |                    ------------- ^------
   |                    |             |
   |  __________________|_____________within this `{closure@$DIR/tests/ui-string/03-guard-is-not-send.rs:9:32: 9:39}`
   | |                  |
   | |                  required by a bound introduced by this call
...  |
14 | |     });
   | |_____^ `*mut String` cannot be sent between threads safely
   |
   = help: within `{closure@$DIR/tests/ui-string/03-guard-is-not-send.rs:9:32: 9:39}`, the trait `Send` is not implemented for `*mut String`
note: required because it appears within the type `ReusableStringGuard<'_>`
  --> src/string.rs
   |
   | pub struct ReusableStringGuard<'parent> {
   |            ^^^^^^^^^^^^^^^^^^^
note: required because it's used within this closure
  --> tests/ui-string/03-guard-is-not-send.rs:9:32
   |
9  |     let handle = thread::spawn(move || {
   |                                ^^^^^^^
note: required by a bound in `spawn`
  --> $RUST/std/src/thread/mod.rs
   |
   | pub fn spawn<F, T>(f: F) -> JoinHandle<T>
   |        ----- required by a bound in this function
...
   |     F: Send + 'static,
   |        ^^^^ required by this bound in `spawn`
